src/phoenix/trace/otel/decode.py

```diff
@@ -0,0 +1,203 @@
+from dataclasses import dataclass, field
+from typing import Any, Container, Dict, Iterable, List, NamedTuple, Optional, Tuple, cast
+from uuid import UUID
+
+import opentelemetry.proto.trace.v1.trace_pb2 as otlp
+from typing_extensions import TypeAlias
+
+from phoenix.trace.otel.utils import (
+    decode_key_values,
+    decode_unix_nano,
+)
+from phoenix.trace.schemas import (
+    Span,
+    SpanContext,
+    SpanEvent,
+    SpanException,
+    SpanID,
+    SpanKind,
+    SpanStatusCode,
+    TraceID,
+)
+from phoenix.trace.semantic_conventions import (
+    DOCUMENT_METADATA,
+    EMBEDDING_EMBEDDINGS,
+    EXCEPTION_ESCAPED,
+    EXCEPTION_MESSAGE,
+    EXCEPTION_STACKTRACE,
+    EXCEPTION_TYPE,
+    LLM_INPUT_MESSAGES,
+    LLM_OUTPUT_MESSAGES,
+    LLM_PROMPT_TEMPLATE_VARIABLES,
+    PHOENIX_EMBEDDING_OBJECT,
+    PHOENIX_RETRIEVED_DOCUMENT,
+    PHOENIX_SPAN_KIND,
+    RETRIEVAL_DOCUMENTS,
+)
+
+
+def decode(otlp_span: otlp.Span) -> Span:
+    trace_id = cast(TraceID, _decode_identifier(otlp_span.trace_id))
+    span_id = cast(SpanID, _decode_identifier(otlp_span.span_id))
+    parent_id = _decode_identifier(otlp_span.parent_span_id)
+
+    start_time = decode_unix_nano(otlp_span.start_time_unix_nano)
+    end_time = (
+        decode_unix_nano(otlp_span.end_time_unix_nano) if otlp_span.end_time_unix_nano else None
+    )
+
+    attributes = dict(decode_key_values(otlp_span.attributes))
+    span_kind = SpanKind(attributes.pop(PHOENIX_SPAN_KIND, None))
+
+    (
+        events,
+        documents,
+        embeddings,
+        input_messages,
+        output_messages,
+        prompt_template_variables,
+    ) = _decode_events(otlp_span.events)
```

Comment:
Discussed offline to keep these in attributes but indexed by location, e.g. `llm.input.messages.0.message`

Reply:
One trade-off of the flattened approach is that it makes the coding more complicated, because without a physical manifestation, nothing prevents a span from entering the system with the following attributes.
```python
attributes = {
    "llm.input.messages.99999999999999999999999999999999.message": ...
}
```